pkg/shipper/gcp.go

```diff
@@ -0,0 +1,155 @@
+package shipper
+
+import (
+	"context"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/api/iterator"
+)
+
+// GCSRemote implements a remote for Google Cloud Storage.
+type GCSRemote struct {
+	logger  log.Logger
+	metrics *gcsRemoteMetrics
+	bucket  *storage.BucketHandle
+}
+
+type gcsRemoteMetrics struct {
+	dirSyncs        prometheus.Counter
+	dirSyncFailures prometheus.Counter
+	uploads         prometheus.Counter
+	uploadFailures  prometheus.Counter
+}
+
+func newGCSRemoteMetrics(r prometheus.Registerer) *gcsRemoteMetrics {
+	var m gcsRemoteMetrics
+
+	m.dirSyncs = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "promlts_gcs_remote_dir_syncs_total",
+		Help: "Total dir sync attempts",
+	})
+	m.dirSyncFailures = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "promlts_gcs_remote_dir_sync_failures_total",
+		Help: "Total number of failed  dir syncs",
+	})
+	m.uploads = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "promlts_gcs_remote_uploads_total",
+		Help: "Total object upload attempts",
+	})
+	m.uploadFailures = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "promlts_gcs_remote_upload_failures_total",
+		Help: "Total number of failed object uploads",
+	})
+
+	if r != nil {
+		prometheus.MustRegister()
+	}
+	return &m
+}
+
+// NewGCSRemote returns a new GCSRemote.
+func NewGCSRemote(logger log.Logger, metricReg prometheus.Registerer, bucket *storage.BucketHandle) *GCSRemote {
+	return &GCSRemote{
+		logger:  logger,
+		bucket:  bucket,
+		metrics: newGCSRemoteMetrics(metricReg),
+	}
+}
+
+func (r *GCSRemote) listDir(ctx context.Context, dir string) *storage.ObjectIterator {
+	return r.bucket.Objects(ctx, &storage.Query{
+		Delimiter: "/",
+		Prefix:    dir,
+	})
+}
+
+// Exists checks if the given directory exists at the remote site.
+func (r *GCSRemote) Exists(ctx context.Context, dir string) (bool, error) {
+	objs := r.listDir(ctx, dir)
+	for {
+		if _, err := objs.Next(); err == iterator.Done {
+			break
+		} else if err != nil {
+			return false, err
+		}
+		// The first object found with the given filter indicates that the directory exists.
+		// XXX(fabxc): do a full check whether the files add up as well?
```

Comment:
what do you mean `add up`? add up to what (:

Reply:
You mean if are files are there?

Reply:
Yea, basically just a bit more detailed check whether things are in sync.

Reply:
Well I think we can skip that now, it will only increase GCS operations but no one actually can write to this after upload now. Only when upload + dir delete was broken. Let's see how often that happens